cpp/src/groupby/sort/group_tdigest.cu

```diff
@@ -0,0 +1,849 @@
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/concatenate.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/merge.cuh>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/detail/tdigest/tdigest.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/utilities/span.hpp>
+
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/structs/structs_column_view.hpp>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/iterator/discard_iterator.h>
+
+namespace cudf {
+namespace groupby {
+namespace detail {
+
+namespace {
+
+// the most representative point within a cluster of similar
+// values. {mean, weight}
+// NOTE: Using a tuple here instead of a struct to take advantage of
+// thrust zip iterators for output.
+using centroid = thrust::tuple<double, double, bool>;
+
+// make a centroid from a scalar with a weight of 1.
+template <typename T>
+struct make_centroid {
+  column_device_view const col;
+
+  centroid operator() __device__(size_type index)
+  {
+    return {static_cast<double>(col.element<T>(index)), 1, col.is_valid(index)};
+  }
+};
+
+// make a centroid from an input stream of mean/weight values.
+struct make_weighted_centroid {
+  double const* mean;
+  double const* weight;
+
+  centroid operator() __device__(size_type index) { return {mean[index], weight[index], true}; }
+};
+
+// merge two centroids
+struct merge_centroids {
+  centroid operator() __device__(centroid const& lhs, centroid const& rhs)
+  {
+    bool const lhs_valid = thrust::get<2>(lhs);
+    bool const rhs_valid = thrust::get<2>(rhs);
+    if (!lhs_valid && !rhs_valid) { return {0, 0, false}; }
+    if (!lhs_valid) { return rhs; }
+    if (!rhs_valid) { return lhs; }
+
+    double const lhs_mean   = thrust::get<0>(lhs);
+    double const rhs_mean   = thrust::get<0>(rhs);
+    double const lhs_weight = thrust::get<1>(lhs);
+    double const rhs_weight = thrust::get<1>(rhs);
+    double const new_weight = lhs_weight + rhs_weight;
+    return {(lhs_mean * lhs_weight + rhs_mean * rhs_weight) / new_weight, new_weight, true};
+  }
+};
+
+/**
+ * @brief A functor which returns the nearest cumulative weight in the input stream prior to the
+ * specified next weight limit.
+ *
+ * This functor assumes the weight for all scalars is simply 1. Under this assumption,
+ * the nearest weight that will be <= the next limit is simply the nearest integer < the limit,
+ * which we can get by just taking floor(next_limit).  For example if our next limit is 3.56, the
+ * nearest whole number <= it is floor(3.56) == 3.
+ */
+struct nearest_value_scalar_weights {
+  thrust::pair<double, int> operator() __device__(double next_limit, size_type)
+  {
+    double const f = floor(next_limit);
+    return {f, max(0, static_cast<int>(next_limit) - 1)};
```

Comment:
Is `next_limit` guaranteed to be within the range of values representable by `int`?

Reply:
In this case yes, since `next_limit` will be limited by the max # of input values in this case (scalar inputs) - so the same limit as cudf column sizes.

Reply:
Okay, I was just wary of the double to int conversion.